File: p2p_library/src/main/java/com/tambapps/p2p/file_sharing/task/SharingTask.java
package com.tambapps.p2p.file_sharing.task;

import com.tambapps.p2p.file_sharing.exception.SharingException;
import com.tambapps.p2p.file_sharing.TransferListener;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Task for sharing a file with another peer
 */
public abstract class SharingTask {

    public static final int MAX_PROGRESS = 100;
    protected final TransferListener transferListener;

    SharingTask(TransferListener transferListener) {
        this.transferListener = transferListener;
    }

    private volatile boolean canceled = false;

    void share(int bufferSize, InputStream is, OutputStream os, long totalBytes) throws IOException {
        byte[] buffer = new byte[bufferSize];
        int count;
        int lastProgress = 0;
        long bytesProcessed = 0;
        int progress;
        while ((count = is.read(buffer)) > 0) {
            if (canceled) {
                return;
            }
            bytesProcessed += count;
            os.write(buffer, 0, count);
            progress = (int) Math.min(MAX_PROGRESS - 1, MAX_PROGRESS * bytesProcessed / totalBytes);
            if (progress != lastProgress) {
                lastProgress = progress;
                if (transferListener != null) {
                    transferListener.onProgressUpdate(progress, bytesProcessed, totalBytes);
                }
            }
        }
        if (bytesProcessed != totalBytes) {
            throw new SharingException("Transfer was not properly finished");
        } else if (transferListener != null) {
            transferListener.onProgressUpdate(MAX_PROGRESS, bytesProcessed, totalBytes);
        }
    }


    public void cancel() {
        canceled = true;
    }

    public TransferListener getTransferListener() {
        return transferListener;
    }

    public boolean isCanceled() {
        return canceled;
    }

}

File: p2p_library/src/main/kotlin/com/tambapps/p2p/fandem/task/SharingTask.kt
package com.tambapps.p2p.fandem.task

import com.tambapps.p2p.fandem.exception.SharingException
import com.tambapps.p2p.fandem.listener.TransferListener
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

/**
 * Task for sharing a file with another peer
 */
abstract class SharingTask internal constructor(val transferListener: TransferListener?) {
  @Volatile
  var isCanceled = false
    private set

  @Throws(IOException::class)
  fun share(bufferSize: Int, `is`: InputStream, os: OutputStream, totalBytes: Long) {
    val buffer = ByteArray(bufferSize)
    var count: Int
    var lastProgress = 0
    var bytesProcessed: Long = 0
    var progress: Int
    while (`is`.read(buffer).also { count = it } > 0) {
      if (isCanceled) {
        return
      }
      bytesProcessed += count.toLong()
      os.write(buffer, 0, count)
      progress = Math.min(MAX_PROGRESS - 1.toLong(), MAX_PROGRESS * bytesProcessed / totalBytes).toInt()
      if (progress != lastProgress) {
        lastProgress = progress
        transferListener?.onProgressUpdate(progress, bytesProcessed, totalBytes)
      }
    }
    if (bytesProcessed != totalBytes) {
      throw SharingException("Transfer was not properly finished")
    } else transferListener?.onProgressUpdate(MAX_PROGRESS, bytesProcessed, totalBytes)
  }

  open fun cancel() {
    isCanceled = true
  }

  companion object {
    const val MAX_PROGRESS = 100
  }

}